src/logic/addressBook/utils/v2-migration.ts

```diff
@@ -0,0 +1,130 @@
+    // no name key, safes already migrated
+    return
+  }
+
+  // make address book entries from the safe names & addresses
+  const safeAbEntries: AddressBookState = Object.values(parsedStoredSafes)
+    .filter(({ name }) => name && isValidAddressBookName(name))
+    .map(({ address, name }) => makeAddressBookEntry({ address, name }))
+
+  // remove names from the safes in place
+  Object.values(parsedStoredSafes).forEach((item) => {
+    item.owners = item.owners.map((owner: any) => owner.address)
+    delete item.name
+  })
+  const migratedSafes = parsedStoredSafes as StoredSafes
+
+  const [state] = states
+  const addressBookKey = `${namespace}${namespaceSeparator}${state}`
+  const storedAddressBook = localStorage.getItem(addressBookKey)
+  const addressBookToStore: AddressBookState = storedAddressBook ? JSON.parse(storedAddressBook) : []
+
+  // mutate `addressBookToStore` by adding safes' information
+  safeAbEntries.forEach((entry) => {
+    const safeIndex = getEntryIndex(addressBookToStore, entry)
+
+    if (safeIndex >= 0) {
+      // update AB entry with what was stored in the safe object
+      addressBookToStore[safeIndex] = entry
+    } else {
+      // add the safe entry to the AB
+      addressBookToStore.push(entry)
+    }
+  })
+
+  // store the mutated address book
+  localStorage.setItem(addressBookKey, JSON.stringify(addressBookToStore))
+
+  // update stored safe
+  localStorage.setItem(safesKey, JSON.stringify(migratedSafes))
+  saveSafes(migratedSafes)
+}
+
+/**
+ * Migrates the AddressBook from a per-network to a global storage under the key `ADDRESS_BOOK` in `localStorage`
+ *
+ *  The migrated structure will be `{ address, name, chainId }`
+ *
+ * @note Also, adds `chainId` to every entry in the AddressBook list.
+ */
+const migrateAddressBook = ({ states, namespace, namespaceSeparator }: StorageConfig): void => {
+  const [state] = states
+  const prefix = `v2_${getNetworkName()}`
+  const storageKey = `_immortal|${prefix}__ADDRESS_BOOK_STORAGE_KEY`
+  const newKey = `${namespace}${namespaceSeparator}${state}`
+
+  if (localStorage.getItem(newKey)) {
+    // already migrated
+    return
+  }
+
+  const storedAddressBook = localStorage.getItem(storageKey)
+
+  if (!storedAddressBook) {
+    // nothing to migrate
+    return
+  }
+
+  const parsedAddressBook = JSON.parse(JSON.parse(storedAddressBook as string))
+
+  const migratedAddressBook = (parsedAddressBook as Omit<AddressBookEntry, 'chainId'>[])
+    // exclude those addresses with invalid names and addresses
+    .filter((item) => {
+      return isValidAddressBookName(item.name) && getWeb3().utils.isAddress(item.address)
+    })
+    .map(({ address, ...entry }) =>
+      makeAddressBookEntry({
+        address,
+        ...entry,
+      }),
+    )
+
+  localStorage.setItem(newKey, JSON.stringify(migratedAddressBook))
+
+  // Remove the old Address Book storage
+  localStorage.removeItem(storageKey)
```

Comment:
This removes the old item only from the localStorage but ImmortalDB actually also saves to the IndexDB. It will remain there forever.

Reply:
True, restoring 👍